### ISSUE TITLE
Buffer overflow in `nethuns_flush_netmap`

### ISSUE THREAD

**Issue by RiccardoSagramoni** [larthia/nethuns#8]
Hello, I believe I've discovered a possible _buffer overflow_ bug in the `nethuns_flush_netmap` function. 
I hope this helps and I'm open to discuss the details.

# Problem

In file `src/nethuns/sockets/netmap.c`, at line [381](https://github.com/larthia/nethuns/blob/2795c4c5ddb586e59c934b45b378bae74fd79f1c/src/nethuns/sockets/netmap.c#L381) a new array of `uint32_t` items is allocated with a capacity of `s->p->last_tx_ring - s->p->first_tx_ring + 1`.

```C
prev_tails = alloca((s->p->last_tx_ring - s->p->first_tx_ring + 1) * sizeof(*prev_tails));
```

In the following **for** cycles (line [384](https://github.com/larthia/nethuns/blob/2795c4c5ddb586e59c934b45b378bae74fd79f1c/src/nethuns/sockets/netmap.c#L384) and line [413](https://github.com/larthia/nethuns/blob/2795c4c5ddb586e59c934b45b378bae74fd79f1c/src/nethuns/sockets/netmap.c#L413))  the `prev_tails` array is accessed using the counter variable `i`, which is initialized at `s->p->first_tx_ring`.

```C
// try to push packets marked for transmission
for (i = s->p->first_tx_ring; i <= s->p->last_tx_ring; i++)
{
    ring = NETMAP_TXRING(s->p->nifp, i);
    prev_tails[i] = ring->tail; // <= BUG
    // ...
}
```

```C
// cleanup completed transmissions: for each completed
// netmap slot, mark the corresponding nethuns slot as
// available (inuse <- 0)
for (i = s->p->first_tx_ring; i <= s->p->last_tx_ring; i++)
{
    uint32_t scan, stop;
    ring = NETMAP_TXRING(s->p->nifp, i);

    stop = nm_ring_next(ring, ring->tail);
    for (scan = nm_ring_next(ring, prev_tails[i]); scan != stop; // <= BUG
            scan = nm_ring_next(ring, scan))
    {
         // ...
    }
}
```

The array `prev_tails` should be accessed starting from the position `0`, but instead it's accessed starting from the position `s->p->first_tx_ring`.
If `s->p->first_tx_ring` is larger than `0`, this would eventually lead to a **buffer overflow** when the counter becomes bigger than the allocated size of the array (`s->p->last_tx_ring - s->p->first_tx_ring + 1`).

# Proposed solution

Shift down the index used to access `prev_tails`, so that the first accessed position in the cycles is the position `0` instead of position `s->p->first_tx_ring`.

Line [387](https://github.com/larthia/nethuns/blob/2795c4c5ddb586e59c934b45b378bae74fd79f1c/src/nethuns/sockets/netmap.c#L387)
```C
prev_tails[i - s->p->first_tx_ring] = ring->tail;
```

Line [419](https://github.com/larthia/nethuns/blob/2795c4c5ddb586e59c934b45b378bae74fd79f1c/src/nethuns/sockets/netmap.c#L419)
```C
for (scan = nm_ring_next(ring, prev_tails[i - s->p->first_tx_ring]); scan != stop;
```

**Comment by RiccardoSagramoni** [comment]
@giuseppelettieri perfect, I've opened a new pull request with the proposed fix (#9)

**Comment by RiccardoSagramoni** [comment]
Fix merged, I'll close the issue
